### PR TITLE
[FEATURE][SERVER] Add NegotiationHandler

### PR DESCRIPTION
**Pull request by Drakulix**
Adds an `INegotiationHandler` to the Server, required to participate in sessions.

### CODE REVIEW

File: de.fu_berlin.inf.dpp.server/src/de/fu_berlin/inf/dpp/server/session/NegotiationHandler.java [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+        IncomingSessionNegotiation negotiation) {
+        negotiation.localCancel("Server does not accept session invites",
+            NegotiationTools.CancelOption.NOTIFY_PEER);
+    }
+
+    @Override
+    public void handleOutgoingProjectNegotiation(
+        AbstractOutgoingProjectNegotiation negotiation) {
+        ProjectNegotiation.Status status = negotiation
+            .run(new ConsoleProgressIndicator(System.out));
```
Comment (saros-infrastructure) [inline de.fu_berlin.inf.dpp.server/src/de/fu_berlin/inf/dpp/server/session/NegotiationHandler.java]: Code Smell Code Smell: Replace this use of System.out or System.err by a logger. ([squid:S106](https://sonarcloud.io/organizations/saros/rules#rule_key=squid%3AS106))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=saros_complete&pullRequest=144&issues=AWSovkpxMDLKieEST687&open=AWSovkpxMDLKieEST687)

[ISSUE_KEY]: AWSovkpxMDLKieEST687

File: de.fu_berlin.inf.dpp.server/src/de/fu_berlin/inf/dpp/server/session/NegotiationHandler.java [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+    public NegotiationHandler(ISarosSessionManager sessionManager) {
+        sessionManager.setNegotiationHandler(this);
+        this.sessionManager = sessionManager;
+    }
+
+    @Override
+    public void handleOutgoingSessionNegotiation(
+        OutgoingSessionNegotiation negotiation) {
+        SessionNegotiation.Status status = negotiation
+            .start(new ConsoleProgressIndicator(System.out));
```
Comment (saros-infrastructure) [inline de.fu_berlin.inf.dpp.server/src/de/fu_berlin/inf/dpp/server/session/NegotiationHandler.java]: Code Smell Code Smell: Replace this use of System.out or System.err by a logger. ([squid:S106](https://sonarcloud.io/organizations/saros/rules#rule_key=squid%3AS106))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=saros_complete&pullRequest=144&issues=AWSovkpxMDLKieEST686&open=AWSovkpxMDLKieEST686)

[ISSUE_KEY]: AWSovkpxMDLKieEST686

Comment (saros-infrastructure) [inline de.fu_berlin.inf.dpp.server/src/de/fu_berlin/inf/dpp/server/session/NegotiationHandler.java]: Code Smell Code Smell: Replace this use of System.out or System.err by a logger. ([squid:S106](https://sonarcloud.io/organizations/saros/rules#rule_key=squid%3AS106))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=saros_complete&pullRequest=144&issues=AWSovkpxMDLKieEST686&open=AWSovkpxMDLKieEST686)

[ISSUE_KEY]: AWSovkpxMDLKieEST686

File: de.fu_berlin.inf.dpp.server/src/de/fu_berlin/inf/dpp/server/session/NegotiationHandler.java (hunk 1 above)
Comment (saros-infrastructure) [inline de.fu_berlin.inf.dpp.server/src/de/fu_berlin/inf/dpp/server/session/NegotiationHandler.java]: Code Smell Code Smell: Replace this use of System.out or System.err by a logger. ([squid:S106](https://sonarcloud.io/organizations/saros/rules#rule_key=squid%3AS106))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=saros_complete&pullRequest=144&issues=AWSovkpxMDLKieEST687&open=AWSovkpxMDLKieEST687)

[ISSUE_KEY]: AWSovkpxMDLKieEST687

Comment (saros-infrastructure) [inline de.fu_berlin.inf.dpp.server/src/de/fu_berlin/inf/dpp/server/session/NegotiationHandler.java]: Code Smell Code Smell: Replace this use of System.out or System.err by a logger. ([squid:S106](https://sonarcloud.io/organizations/saros/rules#rule_key=squid%3AS106))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=saros_complete&pullRequest=144&issues=AWSovkpxMDLKieEST687&open=AWSovkpxMDLKieEST687)

[ISSUE_KEY]: AWSovkpxMDLKieEST687

File: de.fu_berlin.inf.dpp.server/src/de/fu_berlin/inf/dpp/server/session/NegotiationHandler.java (hunk 2 above)
Comment (saros-infrastructure) [inline de.fu_berlin.inf.dpp.server/src/de/fu_berlin/inf/dpp/server/session/NegotiationHandler.java]: Code Smell Code Smell: Replace this use of System.out or System.err by a logger. ([squid:S106](https://sonarcloud.io/organizations/saros/rules#rule_key=squid%3AS106))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=saros_complete&pullRequest=144&issues=AWSovkpxMDLKieEST686&open=AWSovkpxMDLKieEST686)

[ISSUE_KEY]: AWSovkpxMDLKieEST686

Comment (saros-infrastructure) [inline de.fu_berlin.inf.dpp.server/src/de/fu_berlin/inf/dpp/server/session/NegotiationHandler.java]: Code Smell Code Smell: Replace this use of System.out or System.err by a logger. ([squid:S106](https://sonarcloud.io/organizations/saros/rules#rule_key=squid%3AS106))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=saros_complete&pullRequest=144&issues=AWSovkpxMDLKieEST686&open=AWSovkpxMDLKieEST686)

[ISSUE_KEY]: AWSovkpxMDLKieEST686

File: de.fu_berlin.inf.dpp.server/src/de/fu_berlin/inf/dpp/server/session/NegotiationHandler.java (hunk 1 above)
Comment (saros-infrastructure) [inline de.fu_berlin.inf.dpp.server/src/de/fu_berlin/inf/dpp/server/session/NegotiationHandler.java]: Code Smell Code Smell: Replace this use of System.out or System.err by a logger. ([squid:S106](https://sonarcloud.io/organizations/saros/rules#rule_key=squid%3AS106))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=saros_complete&pullRequest=144&issues=AWSovkpxMDLKieEST687&open=AWSovkpxMDLKieEST687)

[ISSUE_KEY]: AWSovkpxMDLKieEST687

File: de.fu_berlin.inf.dpp.server/src/de/fu_berlin/inf/dpp/server/session/NegotiationHandler.java [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+import de.fu_berlin.inf.dpp.negotiation.AbstractOutgoingProjectNegotiation;
+import de.fu_berlin.inf.dpp.negotiation.AbstractIncomingProjectNegotiation;
+import de.fu_berlin.inf.dpp.negotiation.NegotiationTools;
+import de.fu_berlin.inf.dpp.server.progress.ConsoleProgressIndicator;
+import de.fu_berlin.inf.dpp.session.INegotiationHandler;
+import de.fu_berlin.inf.dpp.session.ISarosSessionManager;
+
+public class NegotiationHandler implements INegotiationHandler {
+
+    private static final Logger LOG = Logger
```
[the remark this answers is not in the document]
Comment (Drakulix) [reply de.fu_berlin.inf.dpp.server/src/de/fu_berlin/inf/dpp/server/session/NegotiationHandler.java]: I have just replicated the style that is used through the whole project for almost every logger. I am not sure, I should change that here, because of consistency reasons.

File: de.fu_berlin.inf.dpp.server/src/de/fu_berlin/inf/dpp/server/session/NegotiationHandler.java (hunk 2 above)
Comment (saros-infrastructure) [inline de.fu_berlin.inf.dpp.server/src/de/fu_berlin/inf/dpp/server/session/NegotiationHandler.java]: Code Smell Code Smell: Replace this use of System.out or System.err by a logger. ([squid:S106](https://sonarcloud.io/organizations/saros/rules#rule_key=squid%3AS106))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=saros_complete&pullRequest=144&issues=AWaBiP31n3WdA7VnF3Xm&open=AWaBiP31n3WdA7VnF3Xm)

[ISSUE_KEY]: AWaBiP31n3WdA7VnF3Xm

File: de.fu_berlin.inf.dpp.server/src/de/fu_berlin/inf/dpp/server/session/NegotiationHandler.java (hunk 1 above)
Comment (saros-infrastructure) [inline de.fu_berlin.inf.dpp.server/src/de/fu_berlin/inf/dpp/server/session/NegotiationHandler.java]: Code Smell Code Smell: Replace this use of System.out or System.err by a logger. ([squid:S106](https://sonarcloud.io/organizations/saros/rules#rule_key=squid%3AS106))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=saros_complete&pullRequest=144&issues=AWaBiP31n3WdA7VnF3Xn&open=AWaBiP31n3WdA7VnF3Xn)

[ISSUE_KEY]: AWaBiP31n3WdA7VnF3Xn

File: de.fu_berlin.inf.dpp.server/src/de/fu_berlin/inf/dpp/server/session/NegotiationHandler.java (hunk 2 above)
Comment (saros-infrastructure) [inline de.fu_berlin.inf.dpp.server/src/de/fu_berlin/inf/dpp/server/session/NegotiationHandler.java]: Code Smell Code Smell: Replace this use of System.out or System.err by a logger. ([squid:S106](https://sonarcloud.io/organizations/saros/rules#rule_key=squid%3AS106))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=saros_complete&pullRequest=144&issues=AWaBiP31n3WdA7VnF3Xm&open=AWaBiP31n3WdA7VnF3Xm)

[ISSUE_KEY]: AWaBiP31n3WdA7VnF3Xm

File: de.fu_berlin.inf.dpp.server/src/de/fu_berlin/inf/dpp/server/session/NegotiationHandler.java (hunk 1 above)
Comment (saros-infrastructure) [inline de.fu_berlin.inf.dpp.server/src/de/fu_berlin/inf/dpp/server/session/NegotiationHandler.java]: Code Smell Code Smell: Replace this use of System.out or System.err by a logger. ([squid:S106](https://sonarcloud.io/organizations/saros/rules#rule_key=squid%3AS106))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=saros_complete&pullRequest=144&issues=AWaBiP31n3WdA7VnF3Xn&open=AWaBiP31n3WdA7VnF3Xn)

[ISSUE_KEY]: AWaBiP31n3WdA7VnF3Xn

File: de.fu_berlin.inf.dpp.server/src/de/fu_berlin/inf/dpp/server/session/NegotiationHandler.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (m273d15) [reply de.fu_berlin.inf.dpp.server/src/de/fu_berlin/inf/dpp/server/session/NegotiationHandler.java]: This a refactor TODO of the project. We have to rename the loggers from upper-case to lower-case. Therefore the common style in the project can be misleading :sweat_smile: It would be nice if new loggers already use the lower-case naming.

[the remark this answers is not in the document]
Comment (Drakulix) [reply de.fu_berlin.inf.dpp.server/src/de/fu_berlin/inf/dpp/server/session/NegotiationHandler.java]: Ah good to know. I will change it.

**Comment by Drakulix** [comment]
@m273d15 Fixed, anything else?

File: de.fu_berlin.inf.dpp.server/src/de/fu_berlin/inf/dpp/server/session/NegotiationHandler.java [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+    }
+
+    @Override
+    public void handleOutgoingSessionNegotiation(
+        OutgoingSessionNegotiation negotiation) {
+        SessionNegotiation.Status status = negotiation
+            .start(new ConsoleProgressIndicator(System.out));
+
+        if (status == Status.OK) sessionManager.startSharingProjects(negotiation.getPeer());
+        if (status == Status.ERROR || status == Status.REMOTE_ERROR) log.error(negotiation.getErrorMessage());
```
Comment (m273d15) [inline de.fu_berlin.inf.dpp.server/src/de/fu_berlin/inf/dpp/server/session/NegotiationHandler.java]: could be `else if`

Comment (Drakulix) [reply de.fu_berlin.inf.dpp.server/src/de/fu_berlin/inf/dpp/server/session/NegotiationHandler.java]: Fixed

**Comment by Drakulix** [comment]
@srossbach Yes, I did test this and it was working just fine. However I agree, that this was a huge oversight. I have changed the NegotiationHandler to use an Executor.

I am not sure, if it makes sense for the server to be able to handle concurrent negotiations, so I have used a SingleThreadedExecutor for now, processing only one negotiation at the time (still allowing to queue more). What is your opinion on this subject, should I use a thread pool executor or something similar?

**Comment by srossbach** [comment]
The reason I asked if you had test your changes based on my assumption that this should not work in any ways with the patch set. The packet is received by our packet dispatch thread. The logic now calls your handler. As the negotiation stuff requires exchanging packets you should now only be able to send packets but not be able to receive them as the negotiation is executed by the dispatch thread.

I am still wondering why this works anyway.

For your other question. I recommend to use a cached ThreadPoolExcecutor.

See https://github.com/saros-project/saros/blob/master/de.fu_berlin.inf.dpp.core/src/de/fu_berlin/inf/dpp/net/stream/Socks5StreamService.java Line 697 (sorry I am on a shitty computer and cannot post the direct link), also https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/Executors.html#newCachedThreadPool()

I do not now if you can set a Maximum size but I guess a Maximum of 4 concurrent threads should be fine. At least for the session negotiation. The Project/file negotiation stuff is another Story. As this really depends on the underlying disk System regarding Performance etc.

**Comment by Drakulix** [comment]
> I am still wondering why this works anyway.

I have no answer either, but I am 100% certain, that I did even demo exactly this patch at some point.

> For your other question. I recommend to use a cached ThreadPoolExcecutor.

Done

> I do not now if you can set a Maximum size but I guess a Maximum of 4 concurrent threads should be fine. At least for the session negotiation. The Project/file negotiation stuff is another Story. As this really depends on the underlying disk System regarding Performance etc.

It is not possible to set a maximum, for that case I would need use a FixedThreadPool. The CachedExecutor will always create as many threads as a required for all tasks to run concurrently. This has the potential to DDOS the server, but I am certain there are many was to do that, as the protocol trusts the clients quite much. As a result there is also no point in using different executors for Session- and ProjectNegotiations.

**Comment by srossbach** [comment]
Take a look at https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/ThreadPoolExecutor.html

It is possible to create such a pool (min thread = 0, max thread = n)

File: de.fu_berlin.inf.dpp.server/src/de/fu_berlin/inf/dpp/server/session/NegotiationHandler.java [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+        final OutgoingSessionNegotiation negotiation) {
+
+        executor.execute(new Runnable() {
+            @Override
+            public void run() {
+                SessionNegotiation.Status status = negotiation
+                    .start(new ConsoleProgressIndicator(System.out));
+
+                if (status == Status.OK) sessionManager.startSharingProjects(negotiation.getPeer());
+                else if (status == Status.ERROR || status == Status.REMOTE_ERROR) log.error(negotiation.getErrorMessage());
```
Comment (srossbach) [inline de.fu_berlin.inf.dpp.server/src/de/fu_berlin/inf/dpp/server/session/NegotiationHandler.java]: I do not know what information is contained in the getErrorMessage(). I think it will not hurt to have a better/unique log statement for ERROR,REMOTE_ERROR, LOCAL_CANCEL and REMOTE_CANCEL

File: de.fu_berlin.inf.dpp.server/src/de/fu_berlin/inf/dpp/server/session/ServerSessionContextFactory.java [hunk 6]
```diff
@@ -19,5 +19,6 @@ public final void createNonCoreComponents(ISarosSession session,
         container.addComponent(FileActivityExecutor.class);
         container.addComponent(FolderActivityExecutor.class);
         container.addComponent(TextEditActivityExecutor.class);
+        container.addComponent(NegotiationHandler.class);
```
Comment (srossbach) [inline de.fu_berlin.inf.dpp.server/src/de/fu_berlin/inf/dpp/server/session/ServerSessionContextFactory.java]: As you attach this class here I highly suggest to implement Startable or move this handler into application scope.

**Comment by Drakulix** [comment]
> It is possible to create such a pool (min thread = 0, max thread = n)
Done. I have used 4 for the session negotiations and 10 for the project negotiations. I suspect there will be more of those, but we can tweak these values.

> better/unique log statement for ERROR,REMOTE_ERROR, LOCAL_CANCEL and REMOTE_CANCEL
Done for all negotiations we actually care about.

> So I think this have a hard time as a session scope component as you have to run the negotiation even if it got canceled before to notify the remote side.
Not sure I got that one. I am now calling `accept`/`run` after cancelling to actually execute the cancellation, is that what you meant?

> As you attach this class here I highly suggest to implement Startable or move this handler into application scope.
I don't understand, why should it implement `Startable`? It registers itself with the SessionManager in the constructor, nothing else needs to be done, right?

**Comment by srossbach** [comment]
> Not sure I got that one. I am now calling accept/run after cancelling to actually execute the cancellation, is that what you meant?

+

> I don't understand, why should it implement Startable? It registers itself with the SessionManager in the constructor, nothing else needs to be done, right?

You have 2 scopes / context. Application and session. Everything that goes into the session will be instantiated every a session is started. If such a component references itself to a application component it cannot be garbage collected unless you detach yourself.

And as for the question above I highly recommend you to make this a application component for now as that ensures that you will always call the run/start methods on the negotiation objects.


And I guess there is currently no other plan as calling startSession once on the server side.

**Comment by Drakulix** [comment]
> And as for the question above I highly recommend you to make this a application component for now as that ensures that you will always call the run/start methods on the negotiation objects.

Done. Is this PR good to go now?

File: de.fu_berlin.inf.dpp.server/src/de/fu_berlin/inf/dpp/server/session/NegotiationHandler.java [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+import de.fu_berlin.inf.dpp.util.NamedThreadFactory;
+
+public class NegotiationHandler implements INegotiationHandler {
+
+    private static final Logger log = Logger
+        .getLogger(NegotiationHandler.class);
+
+    private final ISarosSessionManager sessionManager;
+    private final ThreadPoolExecutor sessionExecutor = new ThreadPoolExecutor(0, 4, 60, TimeUnit.SECONDS, new LinkedBlockingQueue<Runnable>(), new NamedThreadFactory(
+        "ServerNegotiationHandler-"));
```
Comment (srossbach) [inline de.fu_berlin.inf.dpp.server/src/de/fu_berlin/inf/dpp/server/session/NegotiationHandler.java]: You used the name twice but these are 2 different context. Session Negotiation and Resource/Project Negotiation.

File: de.fu_berlin.inf.dpp.server/src/de/fu_berlin/inf/dpp/server/session/NegotiationHandler.java [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+
+    @Override
+    public void handleOutgoingProjectNegotiation(
+        final AbstractOutgoingProjectNegotiation negotiation) {
+
+        projectExecutor.execute(new Runnable() {
+            @Override
+            public void run() {
+                ProjectNegotiation.Status status = negotiation
+                    .run(new ConsoleProgressIndicator(System.out));
```
Comment (srossbach) [inline de.fu_berlin.inf.dpp.server/src/de/fu_berlin/inf/dpp/server/session/NegotiationHandler.java]: I believe this is debug stuff, so you should add a flag if. Normally this should be called with a null progress monitor or some logging progress monitor that logs at least beginTask and done calls.

File: de.fu_berlin.inf.dpp.server/src/de/fu_berlin/inf/dpp/server/session/NegotiationHandler.java [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+                        log.error("ERROR running project negotiation: "+negotiation.getErrorMessage());
+                        break;
+                    case REMOTE_ERROR:
+                        log.error("REMOTE_ERROR running project negotiation: "+negotiation.getErrorMessage());
+                        break;
+                    case CANCEL:
+                        log.info("Project negotiation was cancelled locally");
+                        break;
+                    case REMOTE_CANCEL:
+                        log.info("Project negotiation was cancelled by remote");
```
Comment (srossbach) [inline de.fu_berlin.inf.dpp.server/src/de/fu_berlin/inf/dpp/server/session/NegotiationHandler.java]: + negotiation.getPeer() ?

File: de.fu_berlin.inf.dpp.server/src/de/fu_berlin/inf/dpp/server/session/NegotiationHandler.java (hunk 8 above)
Comment (Drakulix) [reply de.fu_berlin.inf.dpp.server/src/de/fu_berlin/inf/dpp/server/session/NegotiationHandler.java]: Do we have some rules for what counts as debug and what does not on the server?

The `ConsoleProgressIndicator` is indeed for user output and not debugging as it only works in a terminal and focuses on keeping a short backlog. Since the ProjectNegotiation can actually log quite a bit interesting information, I was assuming a user-friendly output might actually be desired.

Comment (srossbach) [reply de.fu_berlin.inf.dpp.server/src/de/fu_berlin/inf/dpp/server/session/NegotiationHandler.java]: Are you aware that every service / daemon execution normally ends with 2&1 > somelogfile.txt or 2&1 > /dev/null

Nobody is looking at this output in a release environment. There is a reason SonarQube barks because of the System.out usage.

File: de.fu_berlin.inf.dpp.server/src/de/fu_berlin/inf/dpp/server/session/NegotiationHandler.java (hunk 7 above)
Comment (Drakulix) [reply de.fu_berlin.inf.dpp.server/src/de/fu_berlin/inf/dpp/server/session/NegotiationHandler.java]: Yes, a stupid copy&paste error. fixed

File: de.fu_berlin.inf.dpp.server/src/de/fu_berlin/inf/dpp/server/session/NegotiationHandler.java (hunk 8 above)
Comment (Drakulix) [reply de.fu_berlin.inf.dpp.server/src/de/fu_berlin/inf/dpp/server/session/NegotiationHandler.java]: Alright, this most likely looks total broken in log files, so I am now using the same flag used to check if the `ServerConsole` should be initialized.

File: de.fu_berlin.inf.dpp.server/src/de/fu_berlin/inf/dpp/server/session/NegotiationHandler.java [hunk 10]
```diff
@@ -0,0 +1,146 @@
...
+                    NegotiationTools.CancelOption.NOTIFY_PEER);
+                negotiation.accept(new NullProgressMonitor());
+            }
+        });
+    }
+
+    @Override
+    public void handleOutgoingProjectNegotiation(
+        final AbstractOutgoingProjectNegotiation negotiation) {
+
```
Comment (saros-infrastructure) [inline de.fu_berlin.inf.dpp.server/src/de/fu_berlin/inf/dpp/server/session/NegotiationHandler.java]: Code Smell Code Smell: Complete cases by adding the missing enum constants or add a default case to this switch. ([squid:SwitchLastCaseIsDefaultCheck](https://sonarcloud.io/organizations/saros/rules#rule_key=squid%3ASwitchLastCaseIsDefaultCheck))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=saros_complete&pullRequest=144&issues=AWbKFPs6O08i52vdg2B1&open=AWbKFPs6O08i52vdg2B1)

[ISSUE_KEY]: AWbKFPs6O08i52vdg2B1

File: de.fu_berlin.inf.dpp.server/src/de/fu_berlin/inf/dpp/server/session/NegotiationHandler.java [hunk 11]
```diff
@@ -0,0 +1,146 @@
...
+            public void run() {
+                negotiation.localCancel("Server does not accept session invites",
+                    NegotiationTools.CancelOption.NOTIFY_PEER);
+                negotiation.accept(new NullProgressMonitor());
+            }
+        });
+    }
+
+    @Override
+    public void handleOutgoingProjectNegotiation(
```
Comment (saros-infrastructure) [inline de.fu_berlin.inf.dpp.server/src/de/fu_berlin/inf/dpp/server/session/NegotiationHandler.java]: Code Smell Code Smell: Replace this use of System.out or System.err by a logger. ([squid:S106](https://sonarcloud.io/organizations/saros/rules#rule_key=squid%3AS106))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=saros_complete&pullRequest=144&issues=AWaBiP31n3WdA7VnF3Xn&open=AWaBiP31n3WdA7VnF3Xn)

[ISSUE_KEY]: AWaBiP31n3WdA7VnF3Xn

File: de.fu_berlin.inf.dpp.server/src/de/fu_berlin/inf/dpp/server/session/NegotiationHandler.java [hunk 12]
```diff
@@ -0,0 +1,146 @@
...
+    }
+
+    @Override
+    public void handleOutgoingSessionNegotiation(
+        final OutgoingSessionNegotiation negotiation) {
+
+        sessionExecutor.execute(new Runnable() {
+            @Override
+            public void run() {
+                SessionNegotiation.Status status;
```
Comment (saros-infrastructure) [inline de.fu_berlin.inf.dpp.server/src/de/fu_berlin/inf/dpp/server/session/NegotiationHandler.java]: Code Smell Code Smell: Replace this use of System.out or System.err by a logger. ([squid:S106](https://sonarcloud.io/organizations/saros/rules#rule_key=squid%3AS106))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=saros_complete&pullRequest=144&issues=AWaBiP31n3WdA7VnF3Xm&open=AWaBiP31n3WdA7VnF3Xm)

[ISSUE_KEY]: AWaBiP31n3WdA7VnF3Xm

File: de.fu_berlin.inf.dpp.server/src/de/fu_berlin/inf/dpp/server/session/NegotiationHandler.java (hunk 9 above)
Comment (Drakulix) [reply de.fu_berlin.inf.dpp.server/src/de/fu_berlin/inf/dpp/server/session/NegotiationHandler.java]: Fixed

File: de.fu_berlin.inf.dpp.server/src/de/fu_berlin/inf/dpp/server/session/NegotiationHandler.java [hunk 13]
```diff
@@ -0,0 +1,146 @@
...
+                ProjectNegotiation.Status status;
+                if (ServerConfig.isInteractive()) {
+                    status = negotiation
+                        .run(new ConsoleProgressIndicator(System.out));
+                } else {
+                    status = negotiation
+                        .run(new NullProgressMonitor());
+                }
+                
+                switch (status) {
```
Comment (saros-infrastructure) [inline de.fu_berlin.inf.dpp.server/src/de/fu_berlin/inf/dpp/server/session/NegotiationHandler.java]: Code Smell Code Smell: Complete cases by adding the missing enum constants or add a default case to this switch. ([squid:SwitchLastCaseIsDefaultCheck](https://sonarcloud.io/organizations/saros/rules#rule_key=squid%3ASwitchLastCaseIsDefaultCheck))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=saros_complete&pullRequest=144&issues=AWbKFPs6O08i52vdg2B1&open=AWbKFPs6O08i52vdg2B1)

[ISSUE_KEY]: AWbKFPs6O08i52vdg2B1

File: de.fu_berlin.inf.dpp.server/src/de/fu_berlin/inf/dpp/server/session/NegotiationHandler.java [hunk 14]
```diff
@@ -0,0 +1,146 @@
...
+    public void handleOutgoingSessionNegotiation(
+        final OutgoingSessionNegotiation negotiation) {
+
+        sessionExecutor.execute(new Runnable() {
+            @Override
+            public void run() {
+                SessionNegotiation.Status status;
+                if (ServerConfig.isInteractive()) {
+                    status = negotiation
+                        .start(new ConsoleProgressIndicator(System.out));
```
Comment (saros-infrastructure) [inline de.fu_berlin.inf.dpp.server/src/de/fu_berlin/inf/dpp/server/session/NegotiationHandler.java]: Code Smell Code Smell: Replace this use of System.out or System.err by a logger. ([squid:S106](https://sonarcloud.io/organizations/saros/rules#rule_key=squid%3AS106))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=saros_complete&pullRequest=144&issues=AWaBiP31n3WdA7VnF3Xm&open=AWaBiP31n3WdA7VnF3Xm)

[ISSUE_KEY]: AWaBiP31n3WdA7VnF3Xm

File: de.fu_berlin.inf.dpp.server/src/de/fu_berlin/inf/dpp/server/session/NegotiationHandler.java [hunk 15]
```diff
@@ -0,0 +1,146 @@
...
+    public void handleOutgoingProjectNegotiation(
+        final AbstractOutgoingProjectNegotiation negotiation) {
+
+        projectExecutor.execute(new Runnable() {
+            @Override
+            public void run() {
+                ProjectNegotiation.Status status;
+                if (ServerConfig.isInteractive()) {
+                    status = negotiation
+                        .run(new ConsoleProgressIndicator(System.out));
```
Comment (saros-infrastructure) [inline de.fu_berlin.inf.dpp.server/src/de/fu_berlin/inf/dpp/server/session/NegotiationHandler.java]: Code Smell Code Smell: Replace this use of System.out or System.err by a logger. ([squid:S106](https://sonarcloud.io/organizations/saros/rules#rule_key=squid%3AS106))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=saros_complete&pullRequest=144&issues=AWaBiP31n3WdA7VnF3Xn&open=AWaBiP31n3WdA7VnF3Xn)

[ISSUE_KEY]: AWaBiP31n3WdA7VnF3Xn

File: de.fu_berlin.inf.dpp.server/src/de/fu_berlin/inf/dpp/server/session/NegotiationHandler.java [hunk 16]
```diff
@@ -0,0 +1,146 @@
...
+                    status = negotiation
+                        .start(new NullProgressMonitor());
+                }
+
+                switch (status) {
+                    case OK:
+                        sessionManager.startSharingProjects(negotiation.getPeer());
+                        break;
+                    case ERROR:
+                        log.error("ERROR running session negotiation: "+negotiation.getErrorMessage());
```
Comment (srossbach) [inline de.fu_berlin.inf.dpp.server/src/de/fu_berlin/inf/dpp/server/session/NegotiationHandler.java]: FYI if you write

String s = a + b + c 

This is only syntactic sugar and the compiler will transform this into

String s = new StringBuilder().append(a).append(b).append(c).toString();

where append() will invoke the toString() method on the object you are passing in

TLDR: it is not necessary to call the toString() method as it is done automatically.
